purchase_discount/models/purchase_order.py

```diff
@@ -1,34 +1,23 @@
-# -*- encoding: utf-8 -*-
-##############################################################################
-#
-#    OpenERP, Open Source Management Solution
-#    Copyright (C) 2004-2009 Tiny SPRL (<http://tiny.be>).
-#
-#    This program is free software: you can redistribute it and/or modify
-#    it under the terms of the GNU Affero General Public License as
-#    published by the Free Software Foundation, either version 3 of the
-#    License, or (at your option) any later version.
-#
-#    This program is distributed in the hope that it will be useful,
-#    but WITHOUT ANY WARRANTY; without even the implied warranty of
-#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-#    GNU Affero General Public License for more details.
-#
-#    You should have received a copy of the GNU Affero General Public License
-#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
-#
-##############################################################################
+# -*- coding: utf-8 -*-
+# © 2004-2009 Tiny SPRL (<http://tiny.be>).
+# © 2015 Pedro M. Baeza
+# © 2016 ACSONE SA/NV (<http://acsone.eu>)
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
 from openerp import models, fields, api
 import openerp.addons.decimal_precision as dp
 
 
 class PurchaseOrderLine(models.Model):
     _inherit = "purchase.order.line"
 
-    @api.model
-    def _calc_line_base_price(self, line):
-        res = super(PurchaseOrderLine, self)._calc_line_base_price(line)
-        return res * (1 - line.discount / 100.0)
+    def _calc_line_base_price(self):
+        self.ensure_one()
+        res = super(PurchaseOrderLine, self)._calc_line_base_price()
+        return res * (1 - self.discount / 100.0)
+
+    @api.depends('discount')
+    def _compute_amount(self):
+        super(PurchaseOrderLine, self)._compute_amount()
```

Comment:
As the hooks doesn't seems to be accepted, you can do this:

    @api.depends('discount')
    def _compute_amount(self):
        prices = {}
        for line in self:
            if line.discount:
                prices[line.id] = line.price_unit
                line.price_unit *= (1 - line.discount / 100.0)
        super(PurchaseOrderLine, self)._compute_amount()
        # restore prices
        for line in self:
            if self.discount:
                line.price_unit = prices[line.id]